friend's status route

EtienneDesticourt opened the pull request.

EtienneDesticourt committed:
Start of online_status route implementation

EtienneDesticourt committed:
Fixed player name return, added assert to make sure player is ingame

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -85,11 +87,35 @@ defmodule Entice.Web.Client do
 
+  @doc "Returns a friend's online status and character name from his account id."
+  def get_status(account_id) do
+    case Entice.Web.Repo.get(Entice.Web.Account, account_id) do
```
i think before querying the db we can quickly shoot a query to the client server (below in this source file) - it would need a new dictionary for that purpose, since we now only order by emails, but that should be easy to do...

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -85,11 +87,35 @@ defmodule Entice.Web.Client do
+        acc = Entice.Web.Repo.preload(acc, :characters)
+        case Client.Server.get_client(acc.email) do
+          nil ->
+            if length(acc.characters) == 0 do
```
this if would be better placed in a guard, or maybe we can avoid it with a match

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -85,11 +87,35 @@ defmodule Entice.Web.Client do
+            {:ok, :offline, name}
+          id ->
+            attribute = Entity.fetch_attribute(id, Client)
+            assert attribute != :error, "Friend status should never be fetched before entity token."
```
i dont think asserts in code that is not a test work.
also note that this is not really a valid remark... if we have a client entity but no client attribute, the client is in some kind of intermediate login state, so it's simply `:offline` in this case.
anyway, good catch!

EtienneDesticourt replied:
I tested the assert, it works.
Yeah you're right I made that assert with the thought in mind that we were fetching the route caller's client for some reason, in which case it made no sense for the route to be called before the player is ingame. Since it's another user's client the assert makes no sense, I'll remove it.

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -85,11 +87,35 @@ defmodule Entice.Web.Client do
+            attribute = Entity.fetch_attribute(id, Client)
+            assert attribute != :error, "Friend status should never be fetched before entity token."
+            {:ok, client} = attribute
+            player = Player.attributes(client.entity_id)
```
this looks suspicious to me... i'd probably do it the same way, but i think this will lead to some refactoring in the future (not THIS exact line of code, but the general way of accessing these player attribute doesnt feel very elegant to me... but nevermind that for now)

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -3,12 +3,14 @@ defmodule Entice.Web.Client do
   Stores client related plain data
   """
-  defstruct entity_id: nil
+  defstruct entity_id: nil, online_status: nil
```
we can safely assume a default of `:offline` here

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -16,6 +16,18 @@ defmodule Entice.Web.FriendsController do
+
+    result = case Client.get_status(account_id) do
+      {:error, :no_matching_account} -> error(%{message: "There is no account with that id."})
+      {:ok, status, name} -> ok(%{message: "Friend info recovery successful.", online_status: status, character_name: name})
```
minor thing: is recovery the right word here?

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -16,6 +16,18 @@ defmodule Entice.Web.FriendsController do
+
+    conn |> json result
+  end
 
```
i think we should also be able to reply to batch requests, i.e. if the client sends several account_ids at the same time. otherwise loads of spam is to be expected...
we might even find it too much for what we want to achieve in the future and serve this by means of channels, but thats another story

EtienneDesticourt replied:
I'm unsure how to parse an integer list from conn.params, do you know how :p?

ephe-meral replied:
I think theres a syntax of `...?account_id[]=1&account_id[]=2` or something the like - but I'm not sure if phoenix supports that...

EtienneDesticourt committed:
Added account_id to Agent's map, minor fixes

EtienneDesticourt committed:
Modified status route for batch request, updated test for modified client

EtienneDesticourt committed:
Switch account id from hardcoded to incremented

EtienneDesticourt committed:
Changes to everything friend related to allow querying by name

Added basename to friend model, updated migration file.
Removed status route, replaced by index.
Updated all routes to take a character name.
Added necessary queries.
Updated friends page.

EtienneDesticourt committed:
Implemented friends_controller tests

ephe-meral commented:
please merge develop back in :)
(rebasing is also an option, but you'll get massive conflicts anyway and resolving them in a rebase is not the nicest thing on earth :P)

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -107,12 +131,24 @@ defmodule Entice.Web.Client.Server do
 
-  def set_client(email, client_id) when is_bitstring(email),
-  do: Agent.update(__MODULE__, &Map.put(&1, email, client_id))
+  def set_client(key, client_id) do
```
as discussed in irc... could be eg:

``` Elixir
def set_client_by_account_id(acc_id, client), do: set_client(:acc_id, acc_id, client)
def set_client_by_email(email, client), do: set_client(:email, email, client)

defp set_client(map_key, key, client_id) do
  Agent.update(__MODULE__,
    fn state ->
      new_entry = %{key, client_id}
      Map.update(state, map_key, new_entry, fn map -> Map.merge(map, new_entry))
    end)
end
```

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -107,12 +131,24 @@ defmodule Entice.Web.Client.Server do
 
-  def get_client(email) when is_bitstring(email),
-  do: Agent.get(__MODULE__, &Map.get(&1, email))
+  def get_client(key) do
```
see comment for set_client

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -107,12 +131,24 @@ defmodule Entice.Web.Client.Server do
 
-  def remove_client(email) when is_bitstring(email),
-  do: Agent.update(__MODULE__, &Map.delete(&1, email))
+  def remove_client(key) do
```
see comment for set_client

ephe-meral commented on test/test_factories.exs:
```diff
@@ -26,7 +26,8 @@ defmodule Entice.Test.Factories do
     email: "somemail#{Counter.get_num(:account_email)}@example.com",
-    characters: characters}
+    characters: characters,
+    id: Counter.get_num(:account_id)} #Maybe add incremented id here?
```
remove comment ;)

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -5,36 +5,57 @@ defmodule Entice.Web.FriendsController do
+    friends = for friend <- friends do
+      {:ok, status, name} = get_status(friend.basename)
+      if name != friend.basename do
+        "#{friend.basename} (#{name}) [#{status}]"
```
this is an api route, i think we dont want to render strings here, but rather put it into a json format

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -5,36 +5,57 @@ defmodule Entice.Web.FriendsController do
+      if name != friend.basename do
+        "#{friend.basename} (#{name}) [#{status}]"
+      else
+        "#{friend.basename} [#{status}]"
```
same as above

EtienneDesticourt committed:
Refactored Client.Server functions, changed index route output

EtienneDesticourt committed:
Merge branch 'develop' into friendlist

Conflicts:
	lib/entice/web/client.ex
	priv/repo/migrations/20150919192626_init_friends.exs
	test/support/conn_case.ex
	test/web/controllers/friends_controller_test.exs
	web/controllers/friends_controller.ex
	web/models/friend.ex
	web/models/queries.ex

ephe-meral commented on config/test.exs:
```diff
@@ -16,7 +16,7 @@ config :phoenix, :stacktrace_depth, 20
   adapter: Ecto.Adapters.Postgres,
   username: "postgres",
-  password: "",
+  password: "root",
```
nope, we dont want that here ;) please don't commit local config changes

EtienneDesticourt committed:
get_friend query fix

ephe-meral commented on lib/entice/web/client.ex:
```diff
@@ -111,12 +134,37 @@ defmodule Entice.Web.Client.Server do
-  def remove_client(email) when is_bitstring(email),
-  do: Agent.update(__MODULE__, &Map.delete(&1, email))
+   def get_client(map_key, key) do
+    Agent.get(__MODULE__,
```
maybe:

``` Elixir
Agent.get_and_update(__MODULE__,
  fn %{^map_key => %{^key => value}} = state -> {value, state} 
     %{^map_key => %{}} = state -> {nil, state}
     nil -> {nil, Map.put(%{}, map_key, %{})}
  end)
```

EtienneDesticourt committed:
client fix

EtienneDesticourt committed:
friend_name -> char name

ephe-meral committed:
client.server changed

EtienneDesticourt committed:
Fixed create success test from friends_controller tests

ephe-meral commented on priv/repo/seeds.exs:
```diff
@@ -48,8 +48,8 @@ defmodule Seeds do
-    for {acc_id1, _acc1} <- accounts, {acc_id2, _acc2} <- accounts, acc_id1 != acc_id2,
-    do: Friend.changeset(%Friend{}, %{account_id: acc_id1, friend_account_id: acc_id2}) |> Repo.insert!
+    for {acc_id1, _acc1} <- accounts, {acc_id2, acc2} <- accounts, acc_id1 != acc_id2,
+    do: Friend.changeset(%Friend{}, %{account_id: acc_id1, friend_account_id: acc_id2, base_name: "#{acc2.email} 1"}) |> Repo.insert!
```
Should this be like that? I thought base_name is also a character name? I dont have anything in particular against this though ;)

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -5,34 +5,54 @@ defmodule Entice.Web.FriendsController do
     {:ok, friends} = Entice.Web.Client.get_friends(id)
 
-    friends = for friend <- friends, do: friend.friend_account.id
+    results = []
```
is this necessary?

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -5,34 +5,54 @@ defmodule Entice.Web.FriendsController do
+    results = for friend <- friends do
+      {:ok, status, name} = get_status(friend.base_name)
+      map = %{base_name: friend.base_name, current_name: name, status: status}
+      results = results ++ [map]
```
i dont think we need this if `friends` is a list, which should be the case

ephe-meral replied:
b/c you can then simply return the map you define above

EtienneDesticourt replied:
I was unsure if 'variable = for ...' in elixir worked like list comprehension or if variable was set to the last iteration's return. If I understood what you said correctly, it's the former. I'll make the necessary changes =).

ephe-meral replied:
it's a list comprehension. you can also use Enum.map :)

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -5,34 +5,54 @@ defmodule Entice.Web.FriendsController do
-    friend_account_id = conn.params["friend_account_id"]
-
-    friend_account = Entice.Web.Repo.get(Entice.Web.Account, friend_account_id)
+    friend_name = conn.params["char_name"]
```
we can match on this in the function header. replace `_params` with a `%{"char_name" => friend_name}`

and dont forget to add catch all def for the create function below, so we can handle wrong parameters gracefully. its implemented like this in the other controllers already, check account controller for example

ephe-meral commented on web/controllers/friends_controller.ex:
```diff
@@ -44,9 +64,10 @@ defmodule Entice.Web.FriendsController do
     session_id = get_session(conn, :client_id)
     {:ok, acc} = Client.get_account(session_id)
-    friend_account_id = conn.params["friend_account_id"]
+    friend_name = conn.params["char_name"]
```
same as comment above, match in functionhead

ephe-meral commented on web/models/queries.ex:
```diff
@@ -23,6 +24,16 @@ defmodule Entice.Web.Queries do
+    case get_account_id(name) do
+      {:ok, account_id} ->
+        account = Entice.Web.Repo.get(Entice.Web.Account, account_id)
+        assert account != nil, "There should never be a character without an account." #Right?
```
yes thats true. you can remove the comment ;)

ephe-meral commented on web/models/queries.ex:
```diff
@@ -23,6 +24,16 @@ defmodule Entice.Web.Queries do
     end
   end
 
+  def get_account_by_name(name) do
```
what kind of name are we looking for here? character name?

EtienneDesticourt replied:
Yes, I call the 'def get_account_id(name) do' function in it so I reused the same parameter.

ephe-meral commented on web/models/queries.ex:
```diff
@@ -23,6 +24,16 @@ defmodule Entice.Web.Queries do
   end
 
+  def get_account_by_name(name) do
+    case get_account_id(name) do
```
same as above, what name is this? also this get_account_id function should be renamed to something signifying what the name is...

EtienneDesticourt replied:
I'll rename it to char_name

ephe-meral commented on web/models/queries.ex:
```diff
@@ -75,16 +86,25 @@ defmodule Entice.Web.Queries do
+  def get_friend_by_base_name(account_id, base_name), do: get_friend(account_id, :base_name, base_name)
+  def get_friend_by_friend_account_id(account_id, friend_account_id), do: get_friend(account_id, :friend_account_id, friend_account_id)
+
+  def get_friend(account_id, key_atom, key) do
```
i think we can replace this with Repo.get_by

EtienneDesticourt committed:
Minor refactoring

ephe-meral commented:
please merge or rebase onto develop ;)

EtienneDesticourt committed:
Merge branch 'develop' into friendlist

Conflicts:
	web/controllers/friends_controller.ex

EtienneDesticourt committed:
Minor refactoring
